Public interfaces to manage virt specific methods

sahid opened the pull request:
In the serie we are trying to create clear public interfaces to manage hypervisor and virtual machines.

Even if currently there is no desire of using different virt driver than libvirt. Having a clear interfaces help for code maintainability.

The proposed changes are to introduce virtwrap.Host and virtwrap.Guest, both will create the public interfaces with libvirt client to provide methods for the other components. At this end of this serie it will be prohibited to import anything else from virtwrap than Host and Guest to manage hypervisor connection and virtual machines.

sahid commented:
@berrange will be happy to have some of your thinking on this work in progress

rmohr commented on pkg/virt-handler/virtwrap/host.go:
```diff
@@ -0,0 +1,68 @@
+// will represents the only public solution for the other components
+// to comunicate with internal hypervisor functionalities
+
+type Host interface {
```
Would it make sense to call that all `Hypervisor` instead of `Host`?

sahid replied:
Yes probably, lets use `Hypervisor` if everyone are agreed.

sahid replied:
Well thinking a bit more I do like `Host` because it's a concise name, if you don't have strong discord I would prefer keep it. Sounds good?

rmohr replied:
I don't have a strong opinion here. What I would just keep in mind is, that in the k8s eco-system we have a `Node` which represents a physical or virtual `Host`. We would then also have a `Host` which is our hypervisor driver, which is only as much related to the host, as it runs on the host where virt-handler runs. There are also additions kubevirt native features, regarding to e.g. storage, which are also host local, but not necessarily part of our Host interface (e.g. cloud-init volumes, docker registry images, ...).

In the context of that, I am just not sure if `Host` would really be the wises choice, but I can live with it.

sahid replied:
Fair enough You made that clear `Hypervisor` is what we want.

rmohr commented on pkg/virt-handler/virtwrap/libvirt/libvirt.go:
```diff
@@ -50,6 +49,10 @@ type Connection interface {
+
+	// XXX: This interface is going to be removed to use
+	// virtwrap.Host
+	LookupGuestByName(name string) (VirDomain, error)
```
Right now we have the naming convention of `VirtualMachine` on the cluster and `Domain` on the hypervisor/libvirt level. `Guest`, like you have here, might be  more descriptive than `Domain`. It would also solve the term collision with Domain as "Domain of your application".

rmohr commented:
Could you run `make generate` and commit the result, to make travis happy?

davidvossel commented:
I agree we should strive for public abstractions that are technology agnostic when possible, however it is unclear to me how we benefit from the abstraction being proposed here.

In the end, for this host/guest abstraction to be powerful enough to do what we need it to do, it's going to essentially look just like the libvirt specific bits we have now, right?  At that point, how do we benefit from another layer?  I don't see the possibility of another piece of technology mapping to the host/guest abstraction except for libvirt. 

It is also possible I've missed the intent of what is trying to be achieved here.

rmohr commented:
> I agree we should strive for public abstractions that are technology agnostic when possible, however it is unclear to me how we benefit from the abstraction being proposed here.
> 
> In the end, for this host/guest abstraction to be powerful enough to do what we need it to do, it's going to essentially look just like the libvirt specific bits we have now, right? At that point, how do we benefit from another layer? I don't see the possibility of another piece of technology mapping to the host/guest abstraction except for libvirt.
> 
I agree. It is the libvirt api and it should look like it, since we create our flows and features based on its  capabilities, which is reflected by it's APIs.

Where I would see a potential benefit, is easier to use interfaces by e.g. passing in a struct instead of xml to `dom` operations, automating dom.Free or dom.Close, ... Things like that.

> It is also possible I've missed the intent of what is trying to be achieved here.

rmohr commented:
I would love to have the simplifications I mentioned in the comment above, but we don't have to change the names for that. Everything is already behind interfaces for maintainability and testability. Simplifying the usage, if we see the potential, by still keeping it obvious what this will do in libvirt, is probably a good path.

rmohr commented:
@sahid I don't think that renaming to Guest is necessary. I wonder which simplifications would be possible. 
 Do you have some in mind? 

Could we discuss that before you continue, because like @davidvossel mentioned, just having another interface on the libvirt API matching interfaces does not really help much, if the complexity is equal. Let's make sure we don't waste your time ...

sahid commented:
The current interfaces are just wrapping libvirt-go I don't want to add interfaces of top of them. My desire is to remove them since they provide any high level functionality. The aim of the API I'm trying to provide is to make distinction between low level concepts and the other components which are going to consume them.

A simple example is 'cache.go', it looks like an other low level concept which should not exit. That does not seem right. My desire is to remove it, that by merging it with libvirt package (previously cli).

I'm trying to move incrementally to avoid changing too much code. All of the modules vm, domain, cache, manager, cli are just confusing.

At the end of this work we will have a controller which takes the high level methods to for example initialize a new connection to the host;  which is basically: connect, monitor, register lifecycle events. or create a new VM which is basically: defining spec, create, monitor and moving it to the store.

We don't have lot of code right now so that is why i'm pushing a bit on this way first.

davidvossel commented:
Hey @sahid,

It sounds like you have a decently large refactoring in mind.  I know you've been been involved with VM management for quite some time and what you're striving for here is both useful and probably completely obvious to you. However, the full intent and benefit of these changes are still lost on me.

I'd hate for you to invest too much time in coding this when we're still struggling to share a clear common vision. Would you be willing to create a small (informal) proposal outlining what you hope to achieve and how it benefits the project. We've been submitting proposals in the docs/ directory. There's no formal process for proposals or anything and we don't require it for every change. It's just something I think might benefit us in this discussion.

sahid commented:
@rmohr @davidvossel , Thanks to have jumped on this PR and for your comments. I really appreciate.

I preferred to update the PR to demonstrate where I would like to go. Be sure of that I'm not coming to rework everything without any humility. What you are trying to achieve it's quite impressive and looks really ambitious.

That is said I also understand it's time consuming to discuss of this and I will make my best to make everything concise and easy to review.

In last commit you can see that; `pkg/virt-handler/virtwrap/manager.go` and `pkg/virt-manifest/mapper.go` are defining domain. There are concepts which should from my POV be hidden that because they don't matter how internally libvirt wants the domain defined. I moved everything to 'libvirt.DefineGuestSpec'.

An other point I added TODO in the code. I probably missed some points so feel free to discuss and correct any of them.

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for an this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically.
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organisation's projects.

Note: You might see multiple messages like this showing up on a single PR, this is due to a known bug in the GitHub trigger plugin on Jenkins, the infra team is working on fixing this via new testing flow in the near future, so we'll be able to post a single message per PR.

rmohr commented:
@sahid I am sorry that this did not work out. Thank you for your contribution.
